Add tokenizer for German language

ngulden opened the pull request:
Thanks for the pyenchant project. It helps me a lot with spell check in various Sphinx projects.

With this pull request I suggest a tokenizer for German. The tokenizer is based on the existing tokenizer for English. It adds `-` and `.` to the valid characters list.

With this tokenizer the spell checking works a lot better and saves me some additional work and maintaining some weird dictionaries.

I would love to have this tokenizer included in the project, so I can drop my local implementation as part of a Sphinx extensions.

Thanks.

dmerejkowsky commented:
Since @pmhahn works for a company in Germany, I'm going to ask them to weigh in on this :)

pmhahn commented:
> Since @pmhahn works for a company in Germany, I'm going to ask them to weigh in on this :). The patch looks good to me but I don't speak German.

FYI: To make things explicit: @ngulden is a college of me; we both work for Univention GmbH in Bremen; Actually this German `tokenizer` is based on my work I did 7 years ago for your DocBook spell-checking.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+from typing import Dict, Sequence, Type, Union
+
+from enchant.tokenize import EmailFilter, Filter, URLFilter, basic_tokenize
+from enchant.tokenize import tokenize as base_tokenize  # noqa F401
```
As `pyenchant` is Python-3-only by now, the new type annotation should be used wich is available since Python 3.5 for functions and since 3.6 for variables; Then the `  # noqa F401` is no longer required.
At least it should include the colon: `  # noqa: F401`

ngulden replied:
I removed the `#noqa F401`.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+        super().__init__(text, valid_chars)  # type: ignore
+
+
+TOKENIZERS = {
+    "de": wrap_tokenizer(EmailFilter(URLFilter(basic_tokenize)), tokenize),
+    "": basic_tokenize,
+}  # type: Dict[str, Union[Type[base_tokenize], Filter]]
```
Use Python3 type annotation syntax:
```
TOKENIZERS: Dict[str, Union[Type[base_tokenize], Filter]] = {
    "de": wrap_tokenizer(EmailFilter(URLFilter(basic_tokenize)), tokenize),
    "": basic_tokenize,
}
```

ngulden replied:
I changed it to the Python 3 type annotation as suggested. Thank you.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+from enchant.tokenize.en import tokenize as tokenizer_en
+
+
+class tokenize(tokenizer_en):  # noqa N801
```
Minimum add colon: `  # noqa: F401`

ngulden replied:
I removed the `# noqa N801`.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+        super().__init__(text, valid_chars)  # type: ignore
+
+
+TOKENIZERS = {
```
Where does this come from and for what is this required?

ngulden replied:
What line exactly to do you mean? Line 50 or line 53?

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+
+
+TOKENIZERS = {
+    "de": wrap_tokenizer(EmailFilter(URLFilter(basic_tokenize)), tokenize),
```
Why are these filters always added? This should be the choice of the using party and not hard-coded here.

ngulden replied:
I agree. I thought they must be added here. But looking at how sphinxcontrib.spelling works, that client adds filters itself. So hard coding them here is wrong. I removed that part.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,56 @@
+
+
+class tokenize(tokenizer_en):  # noqa N801
+    def __init__(self, text: str, valid_chars: Sequence[str] = ("-", ".")) -> None:
+        super().__init__(text, valid_chars)  # type: ignore
```
The signature should be changed to be compatible with the super-class, which should make the `# type:ignore` obsolete:
```
    def __init__(self, text: _TextLike, valid_chars: Optional[Container[str]] = ("-", ".")) -> None:
        super().__init__(text, valid_chars)
```

+ `from .en import _Textlike` and `from typing import Container, Optional`

ngulden replied:
I changed the signature as suggested and updated the respective imports.

pmhahn reviewed (changes requested):
Fast review with several questions.

ngulden commented:
Now the tokenizer can't be imported. I investigate it.

ngulden commented:
My environment uses pyenchant version 3.2.2 which doesn't have `_TextLike` defined yet. I found a solution for my local environment. And now the pull request was also updated. I removed the not needed code parts and added the type annotations.

I think, the code can now be reviewed again. Thanks for having a look.

pmhahn reviewed (approved):
Looks fine now. Thank you for this contribution.

pmhahn commented on enchant/tokenize/de.py:
```diff
@@ -0,0 +1,51 @@
+from .en import _TextLike
+
+
+class tokenize(tokenizer_en):
```
```suggestion
class tokenize(tokenizer_en):  # noqa: N801
```

raphaelm commented:
Just as a note, the release of this required us to add *a lot* of stuff to our whitelist: https://github.com/pretix/pretix/pull/5479/files

@ngulden  Should something like "Enterprise-Lizenz" really be considered a typo if "Enterprise" is part of the wordlist and "Lizenz" part of the dictionary? I don't think this maps well how the German language owrks.
